### PR TITLE
fat client implementation

### PR DESCRIPTION
**Pull request by samongoose**

### CODE REVIEW

File: lekko_client/evaluation/evaluation.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
+from dataclasses import dataclass
+from typing import List, Optional
+
+from google.protobuf.any_pb2 import Any
```
Comment (passichenko) [inline lekko_client/evaluation/evaluation.py:4]: nit: maybe it's just me, but when I see `Any` I think that it's `typing.Any`. Maybe import it as `ProtoAny`? Or import package and use it as `any_pb2.Any`?

File: tests/evaulation/test_evaluation.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import pytest
```
Comment (passichenko) [inline tests/evaulation/test_evaluation.py:1]: typo in directory name: `evaulation`

File: lekko_client/evaluation/rules.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+            )
+        else:
+            raise ValueError("unknown comparison operator")
+    elif rule_type == "call_expression":
+        if rule_value.WhichOneof("function") == "bucket":
+            return evaluate_bucket(rule_value.bucket, namespace, config_name, context)
+        else:
+            raise ValueError("unknown function type")
+    else:
+        raise ValueError("unknown rule type")
```
[the remark this answers is not in the document]
Comment (passichenko) [reply lekko_client/evaluation/rules.py:85]: same in other places where we handle proto one-of

File: lekko_client/evaluation/rules.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+
+def evaluate_equals(rule_value: Value, context_value: LekkoValue) -> bool:
+    rule_kind = rule_value.WhichOneof("kind") or ""
+    context_kind = context_value.WhichOneof("kind") or ""
```
[the remark this answers is not in the document]
Comment (samongoose) [reply lekko_client/evaluation/rules.py:90]: to placate mypy

File: lekko_client/evaluation/rules.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        return context_str.startswith(rule_str)
+    elif comparison_operator == ComparisonOperator.COMPARISON_OPERATOR_ENDS_WITH:
+        return context_str.endswith(rule_str)
+    elif comparison_operator == ComparisonOperator.COMPARISON_OPERATOR_CONTAINS:
+        return rule_str in context_str
+    else:
+        raise ValueError("unexpected string comparison operator")
+
+
+def get_string(value: Union[Value, LekkoValue]) -> str:
```
[the remark this answers is not in the document]
Comment (samongoose) [reply lekko_client/evaluation/rules.py:119]: i was trying to, but we can drop support if everyone would prefer

[the remark this answers is not in the document]
Comment (passichenko) [reply lekko_client/evaluation/rules.py:119]: well, looks like some of the dependencies that we are using are 3.5+ anyway :)

File: lekko_client/__init__.py [hunk 6]
```diff
@@ -1,9 +1,89 @@
...
+
+def initialize(
+    mode: Mode,
+    owner_name: str,
+    repo_name: str,
+    api_key: Optional[str] = None,
+    context: Optional[Dict[str, Any]] = None,
+    git_repo_path: Optional[str] = None,
+) -> Client:
+    global __client
```
Comment (passichenko) [inline lekko_client/__init__.py:39]: I don't see it being used, is it for testing? Do we really need this global variable?

Comment (samongoose) [reply lekko_client/__init__.py:39]: i'm open to suggestions on refining the interface, but since we're caching and managing lifecycle, i wanted to make a static interface that has access to the intitialized client. the methods below all make use of `__client` when you call `lekko_client.get_bool` for example. i return the `client` here as a convenience but i don't think it will make sense for sdk users to hold onto it - better to call `lekko_client.initialize()` in their startup function and then just `lekko_client.get_*` everywhere else.

File: lekko_client/stores/store.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        ...
+
+    def load(self, contents: GetRepositoryContentsResponse) -> bool:
+        if not contents:
+            return False
+
+        contents = self.sort_contents(contents)
+        content_hash = self.hash_contents(contents)
+
+        if not self.should_update(contents, content_hash):
```
Comment (passichenko) [inline lekko_client/stores/store.py:26]: Don't we need to have a lock here? Event if updating contents is atomic, commit sha and content_hash could be our of sync.

File: lekko_client/__init__.py (hunk 6 above)
Comment (passichenko) [reply lekko_client/__init__.py:39]: Agree that it will be more convenient for some users. But I have a few concerns with this approach:
- `lekko_client.get_*` relies on user calling `lekko_client.initialize()` beforehand, and it can cause surprise runtime errors after refactoring or reusing a library in some other place. So at least we need to handle it with actionable message.
- it's not compatible with multi-repo setup and lead to unexpected behavior if `lekko.initialize()` is called multiple times. We probably need to have a lock here and throw actionable error if `initialize` is called more than once
- it pushed users to code style that is less composable (how to mix configs form different repo / different modes in one binary) and harder to test (we probably need to provide a helper to mock global client properly anyway)

I'm not against exposing this kind of interface as an option, but I think that we need to handle corner cases that I mentioned above (there are probably more) and clearly document limitations of this approach for users.
I can check with rest of the team if anyone has more opions.

File: lekko_client/stores/store.py (hunk 7 above)
Comment (samongoose) [reply lekko_client/stores/store.py:26]: yeah, that's on the todo list. i should document what's left to do in this pr.

Comment (passichenko) [reply lekko_client/stores/store.py:26]: oh, sorry maybe I jumped to review it too early :( ? then just let me know when it's ready for the next round of review

Comment (samongoose) [reply lekko_client/stores/store.py:26]: sorry - should have been clear. we're not 100% feature complete, but this is ready for feedback on the interface/structure.  We'll be adding more robustness here, plus tests, and the event logging.

File: lekko_client/__init__.py (hunk 6 above)
Comment (DavidSGK) [reply lekko_client/__init__.py:39]: I think personally it makes more sense to allow users to initialize multiple times and users can manage their own global instance of a client if they want. Leaving the instance management to the users is also probably better for supporting patterns such as dependency injection which they may have

Comment (samongoose) [reply lekko_client/__init__.py:39]: my primary concern is that managing their own global instance can be a bit awkward and is susceptible to users inadvertently creating ephemeral instances - which is quite bad considering the two new modes are spinning up background threads to monitor for config updates.

there are ways to manage the instance for them and still make this a bit more DI friendly.

Comment (shubhitms) [reply lekko_client/__init__.py:39]: i don't think we need to solve the multi-repo / multi-mode setup right now. we do need to manage the lifecycle properly and allow users to initialize + destroy the client. 

initialization would involve registration and filling the cache, and
destroy / close would involve deregistration.

i'm not a fan of global variables in general but i think whether or not the underlying implementation uses a global variable should be transparent to the user, as long as the edge cases @passichenko mentioned are handled.

we could expose a dedicated method to initialize the client instead of relying on the class constructor or `initialize()` method - e.g. 

```py
def cached_api_client(...)
...
```

which would invoke the constructor and perform any initialization needed. that way we can avoid issues around having the caller call `initialize()` twice. 

lmk if i'm missing any context on this.

Comment (samongoose) [reply lekko_client/__init__.py:39]: the only issue with the `cached_api_client()` is how cumbersome it would be for all api consumers to pass all initialization parameters just in case the client hasn't been constructed yet. for proper lifecycle management we really should encourage clients to call `initialize()` during their app startup - particularly so they don't incur the cost during request handling

Comment (shubhitms) [reply lekko_client/__init__.py:39]: i guess the key philosophy here is dependency injection. when you have a non-static dependency that's used all across your codebase, what's the best practice for sharing that dependency in python?

in other languages (go, node, rust) we would inject the dependency vs having static global

Comment (passichenko) [reply lekko_client/__init__.py:39]: > what's the best practice for sharing that dependency in python?

I haven't really used python outside of relatively large projects inside a monorepo. In my experience we always end up going with dependency injection (usually manually passing things into constructors)

**Comment by samongoose** [comment]
Quick update - I'm dropping Python3.9 support because it cleans up config and initialization a lot. If someone needs it it won't be too difficult to add back the support.


Still left todo: primarily tests

File: lekko_client/__init__.py [hunk 8]
```diff
@@ -1,9 +1,144 @@
...
+    lekko_uri = LEKKO_SIDECAR_URL
+
+
+@dataclass(kw_only=True)
+class APIConfig(Config):
+    pass
+
+
+@dataclass(kw_only=True)
+class CachedServerConfig(Config):
```
Comment (passichenko) [inline lekko_client/__init__.py:46]: I think that we'll need to have an option to set custom refresh interval here.

File: lekko_client/__init__.py [hunk 9]
```diff
@@ -1,9 +1,144 @@
...
+    with __client_lock:
+        if __client:
+            __client.close()
+        __client = client
+
+
+def __get_safe(func):
+    @functools.wraps(func)
+    def wrapper(*args, **kwargs):
+        with __client_lock:
```
Comment (passichenko) [inline lekko_client/__init__.py:95]: I wonder what are the performance implications of getting a lock for every read (especially if this is non-cached client). AFAIK, python doesn't have rw locks in stdlib and we probably don't want to invent a bicycle here :)
What if we don't grab a lock on reads and declare that `initialize` and `set_client` should be used during initialization only and are not safe to call after the client is in use.

File: lekko_client/__init__.py (hunk 6 above)
Comment (passichenko) [reply lekko_client/__init__.py:85]: Another thing that I don't understand with global instance pattern is how it helps with shutdown and deregistration. Who is responsible for properly closing lekko client?

File: lekko_client/clients/cached_git_client.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+                        sha=sha.decode("utf-8"),
+                        feature=feature,
+                    )
+                )
+        return features
+
+    def close(self):
+        super().close()
+        # if self.events_batcher:
+        # await self.events_batcher.close()
```
Comment (passichenko) [inline lekko_client/clients/cached_git_client.py:109]: old code? parent class should take care of it, right?

**Comment by passichenko** [comment]
I still see some todos, so assuming that it's still work in progress, just left some comments while I was reading

File: lekko_client/__init__.py [hunk 11]
```diff
@@ -1,9 +1,144 @@
 """Lekko Python SDK Client"""
```
Comment (samongoose) [inline lekko_client/__init__.py:1]: missing `close()`

**Comment by DavidSGK** [comment]
One current problem: events are being dropped on client close if the thread is `join`ed before the internal queue is drained fully.

**Comment by passichenko** [comment]
> One current problem: events are being dropped on client close if the thread is joined before the internal queue is drained fully.

Fixed that
